### ISSUE TITLE
Fix for flow upsampling creating super-tiles in extreme dark and 4 other fixes

### ISSUE THREAD

**Issue by Shadetail** [Jamy-L/Handheld-Multi-Frame-Super-Resolution#52]
I was experimenting with version 1.3.0 with some ultra dark photos. I noticed large tile artifacts, which would get worse the darker and more difficult to align the image would be. So naturally the best example of it was by far the worst and darkest burst I had (17 photo burst), this being 2071x1386 (out of 5472x3648) crop of the unprocessed first frame , just to give you an idea:

<img width="2071" height="1386" alt="Image" src="https://github.com/user-attachments/assets/5c3c282d-3a70-4cdd-86c5-df4344c0f784" />

After processing this burst the merged result would come out looking like this, after adjusting exposure etc:
<img width="2071" height="1386" alt="Image" src="https://github.com/user-attachments/assets/db72a696-4504-4e23-9b75-9216806487a1" />

And with the AI denoise in ACR:
<img width="2071" height="1386" alt="Image" src="https://github.com/user-attachments/assets/b38f4ef6-586b-4043-95f6-6fc11eda05be" />

I noticed that those seams are not present in my v1.2.2, and that in some places I was getting slightly better results, but v.1.3.0 had far superior black handling in ultra dark photos for some reason and also handles tiny seams better, so I hunted down what was causing the large seams in 1.3.0 and found and fixed 4 more bugs along the way, again just vibe coding with Claude Code and ChatGPT5 Pro for getting unstuck.

In the end we managed to fix it all and got the output to look like this:

<img width="2071" height="1386" alt="Image" src="https://github.com/user-attachments/assets/f0d08c4d-504b-4342-b455-88cf70badaf9" />

And with the AI denoise in ACR:

<img width="2071" height="1386" alt="Image" src="https://github.com/user-attachments/assets/f0923efb-e3d8-4f87-8569-c52d16eebbef" />

After we were done I asked Claude to explain all the bug fixes so I can share them with you, following is that explanation:

# Bug Fixes for v1.3.0

This document summarizes the bugs found and fixed in v1.3.0 that were causing quality degradation compared to v1.2.2.

## Executive Summary

Five bugs were identified and fixed:
1. SNR-based parameter inversion (minor impact - detail loss)
2. **Flow upsampling creating super-tiles (CRITICAL - primary cause of 256×256 seam artifacts)**
3. Image size mismatch in ICA (minor impact)
4. Missing synchronization between Numba and PyTorch (race condition)
5. Dimension typo in 64×64 ICA kernel (minor impact)

**Result:** After fixes, v1.3.0 produces images with quality exceeding v1.2.2, with all seam artifacts eliminated.

---

## Bug #1: SNR Parameter Inversion ✅ FIXED

**File:** `handheld_super_resolution/params.py`
**Lines:** 84, 88, 92

**Issue:** During refactoring to use `lerp()`, the y_range values were reversed for three parameters, causing noisy images to receive insufficient denoising and clean images to be over-smoothed.

**Fix:**
```python
# Change from:
config.merging.tuning.k_denoise = lerp(SNR, [6, 30], [3.0, 5.0])
config.merging.tuning.D_th = lerp(SNR, [6, 30], [0.71, 0.81])
config.merging.tuning.D_tr = lerp(SNR, [6, 30], [1, 1.24])

# To:
config.merging.tuning.k_denoise = lerp(SNR, [6, 30], [5.0, 3.0])
config.merging.tuning.D_th = lerp(SNR, [6, 30], [0.81, 0.71])
config.merging.tuning.D_tr = lerp(SNR, [6, 30], [1.24, 1])
```

---

## Bug #2: Flow Upsampling Creates Super-Tiles ✅ FIXED (CRITICAL)

**File:** `handheld_super_resolution/alignment.py`
**Function:** `upscale_lvl()`

**Issue:** The most critical bug. Using `torch.repeat_interleave()` to upsample flow fields creates piece-wise constant "super-tiles" where multiple tiles share exactly the same flow vector. With default config (tile_size=32, factors=[1,2,4,4]), this creates 8×8 groups of tiles with identical flow (32 × 8 = 256 pixels), causing visible 256×256 pixel seam grids during merging.

v1.2.2 used content-aware upsampling that selected between neighboring flows based on patch similarity, which broke the super-tile pattern.

**Fix:** Replace `repeat_interleave` with bilinear interpolation:

```python
def upscale_lvl(alignments, npatchs, l, config):
    new_tile_size = config.block_matching.tuning.tile_sizes[l]
    prev_tile_size = config.block_matching.tuning.tile_sizes[l+1]
    upsampling_factor = config.block_matching.tuning.factors[l+1]

    repeat_factor = upsampling_factor // max(1, (new_tile_size // prev_tile_size))

    # Smooth flow upsample avoids piece-wise constant "super-tiles"
    # v1.2.2 used content-aware selection; bilinear is simpler and breaks the super-tile pattern
    flow = alignments.permute(2, 0, 1).unsqueeze(0)   # [H, W, 2] -> [1, 2, H, W]
    upsampled = F.interpolate(flow, scale_factor=repeat_factor,
                              mode='bilinear', align_corners=False)
    upsampled_alignments = upsampled.squeeze(0).permute(1, 2, 0)  # [1, 2, H', W'] -> [H', W', 2]
    upsampled_alignments *= upsampling_factor

    # Add a potential tile with 0 flow on the bottom or the right
    if upsampled_alignments.shape[0] < npatchs[0] or upsampled_alignments.shape[1] < npatchs[1]:
        upsampled_alignments = F.pad(
            upsampled_alignments,
            (0, 0,
             0, npatchs[1] - upsampled_alignments.shape[1],
             0, npatchs[0] - upsampled_alignments.shape[0]),
            mode='constant', value=0)

    return upsampled_alignments
```

**Impact:** This single change eliminates all 256×256 seam artifacts.

---

## Bug #3: Image Size Mismatch in ICA ✅ FIXED

**File:** `handheld_super_resolution/ICA.py`
**Function:** `align_lvl_ica()`

**Issue:** Gaussian pyramid downsampling can create different image sizes for reference vs moving images (e.g., ref: 3680×5504, moving: 3672×5496). The ICA kernel assumes both images have the same size, causing coordinate misalignment.

**Fix:** Add padding to match sizes before ICA:

```python
def align_lvl_ica(ref_img, ref_gradx_lvl, ref_grady_lvl, ref_hessian_lvl,
                  moving_lvl, alignment, l, config):
    verbose_3 = config.verbose >= 3
    tile_size = config.block_matching.tuning.tile_sizes[l]

    # FIX: Ensure moving image matches reference size for coordinate system consistency
    ref_h, ref_w = ref_img.shape
    mov_h, mov_w = moving_lvl.shape
    if mov_h != ref_h or mov_w != ref_w:
        # Pad moving image to match reference size
        import torch.nn.functional as F
        pad_h = ref_h - mov_h
        pad_w = ref_w - mov_w
        # Add batch/channel dims for F.pad, then remove them
        moving_lvl = F.pad(moving_lvl[None, None], (0, pad_w, 0, pad_h), mode='replicate')[0, 0]

    # ... rest of function
```

---

## Bug #4: Missing Numba/PyTorch Synchronization ✅ FIXED

**File:** `handheld_super_resolution/ICA.py`
**Function:** `align_lvl_ica()`

**Issue:** PyTorch uses per-thread default streams while Numba CUDA kernels use the global default stream. Without explicit synchronization, PyTorch can read the `alignment` tensor before the Numba kernel finishes writing, causing race conditions and stale data.

**Fix:** Add explicit synchronization after ICA kernel launch:

```python
cuda_kernel[blockspergrid, threadsperblock](
    ref_img, ref_gradx_lvl, ref_grady_lvl, ref_hessian_lvl,
    moving_lvl, alignment, config.ica.tuning.n_iter)

# *** CRITICAL ***: make Numba->PyTorch visible ordering explicit.
# PyTorch uses per-thread default streams; Numba uses the CUDA default stream.
# Without this barrier, PyTorch can read 'alignment' before the kernel writes finish.
cuda.synchronize()
```

---

## Bug #5: Dimension Typo in 64×64 ICA Kernel ✅ FIXED

**File:** `handheld_super_resolution/ICA.py`
**Function:** `ica_kernel_64()`
**Lines:** ~428-432

**Issue:** The `grady` bounds checks incorrectly used height `h` instead of width `w` for x-coordinate bounds, causing incorrect gradient zeroing when width < height.

**Fix:**
```python
# Change from:
l_grady[1] = grady[y, x + 1] if is_inbound and (x + 1) < h else 0.0
l_grady[2] = grady[y, x + 2] if is_inbound and (x + 2) < h else 0.0
l_grady[3] = grady[y, x + 3] if is_inbound and (x + 3) < h else 0.0

# To:
l_grady[1] = grady[y, x + 1] if is_inbound and (x + 1) < w else 0.0
l_grady[2] = grady[y, x + 2] if is_inbound and (x + 2) < w else 0.0
l_grady[3] = grady[y, x + 3] if is_inbound and (x + 3) < w else 0.0
```

---

## Important Note: ICA Convergence is Normal

During investigation, we discovered that large alignment errors (200-250 pixels) after ICA are **normal and expected**. Testing v1.2.2 revealed nearly identical behavior:

**v1.2.2 (working):**
- Image 1: Before ICA: max=239px → After ICA: max=239px (change=10px)
- Image 2: Before ICA: max=188px → After ICA: max=188px (change=12px)

**v1.3.0 (after fixes):**
- Image 1: max=195px (ICA change=25px)
- Image 2: max=176px (ICA change=15px)

ICA provides only ~10-20 pixel refinements per iteration. Large residual errors are handled by robustness masking and merging. The seam artifacts were **not** caused by insufficient ICA convergence, but by the flow upsampling bug (#2).

## Files Modified

1. `handheld_super_resolution/params.py` - Lines 84, 88, 92 (SNR fix)
2. `handheld_super_resolution/alignment.py` - Function `upscale_lvl()` (critical super-tile fix)
3. `handheld_super_resolution/ICA.py` - Multiple locations:
   - `align_lvl_ica()` - Image padding and synchronization
   - `ica_kernel_64()` - Dimension bounds fix

---

That's the report. I also tested these fixes on the normally exposed images and didn't see any issues with them.

**Comment by Jamy-L** [comment]
Thank you for finding all this !

Bug 1:
Yes definitely a bug!

Bug 2:
This is not a bug per-say but rather an improvement, and I am genuinely surprised to see it has such a big effect: thank you for noticing this. I didn't had burst with very lows SNR and moving subjects, so it may explain why I have never seen this. I have pushed a fix to make the upscale mode configurable between "nearest" (current method), "bilinear" and "bicubic". Maybe bicubic would work even better

Bug 3:
There shouldn't be any case where there is a shape mismatch. The reference and the moving images have the same shape, so the layers of the pyramid must have the same shapes as well. Feel free to let me know/reopen if I missed something here

Bug 4:
This make sense, but I dont have enough understanding of numba and pytorch to confirm it. I have included it to be extra safe

Bug 5:
This is a big bug but this kernel was rewritten in the release 1.3.1, and it appeared to work as expected in my tests


I am legitimately curious whether your improvement really come from the flow upsampling method or from the other bug fixs. If the scene has a lot of movement, I can imagine it making a big difference

**Comment by Shadetail** [comment]
Thanks for such a quick update!
I tested 1.3.2 on this same image, confirming that the nearest is causing those seams. Interestingly I also got a worse result than with 1.3.0, using all the default settings in the config yaml file for both versions (other than flow_upscale_mode of course). Not that it matters compared to 1.3.0 output, but considering just 1.3.2 outputs only I'd say bilinear beats bicubic.
I also ran everything through ACR AI denoise as the results are easier to compare that way:

Single frame:
<img width="2479" height="1656" alt="Image" src="https://github.com/user-attachments/assets/b19614e3-41f0-470b-8681-5b4b0024066f" />

1.3.0 with my fixes:
<img width="2491" height="1664" alt="Image" src="https://github.com/user-attachments/assets/61b34fef-475e-404b-9125-81ee6decbfdd" />

1.3.2 with flow_upscale_mode: nearest:
<img width="2491" height="1664" alt="Image" src="https://github.com/user-attachments/assets/eddd4a71-4fad-48b3-8529-3d3762716218" />

1.3.2 with flow_upscale_mode: bilinear:
<img width="2491" height="1664" alt="Image" src="https://github.com/user-attachments/assets/a9994a41-4cd8-4eef-9050-7f6027e8b808" />

1.3.2 with flow_upscale_mode: bicubic:
<img width="2491" height="1664" alt="Image" src="https://github.com/user-attachments/assets/d7aad387-3c32-445a-aa96-4f5a77915ca4" />

And I uploaded the burst itself to google drive in case you're interested to investigate this further: https://drive.google.com/drive/folders/13kSMSpBrDomNM4QrkrANHcGkqGENcYfG?usp=sharing
Though just to note that I don't consider this burst and the result usable in any case, I would have thrown it away if I wasn't using it to as a stress test for this.

Also Claude (based on 110k tokens of context of our debugging session) left this reply for you regarding your comments on Bug 3 and 4:

---

## Bug 3 (Image Size Mismatch)

I understand your skepticism - in theory, pyramid layers should have matching shapes. However, we did observe this in practice during debugging. Here's the actual debug output we captured:

```
DEBUG ICA level 0: ref_img.shape=torch.Size([3680, 5504]),
                   moving.shape=torch.Size([3680, 5504])
DEBUG ICA level 0: ref_img.shape=torch.Size([3680, 5504]),
                   moving.shape=torch.Size([3672, 5496])
```

The first image had matching shapes, but the second didn't. This was consistent across multiple images in the burst.

**Possible explanations:**
1. Edge case in Gaussian pyramid downsampling with certain image dimensions
2. Difference in how reference vs. moving images are processed before pyramiding
3. May have been inadvertently fixed in your v1.3.1 rewrites

## Bug 4 (Synchronization)

Thank you for including it! The race condition we observed was:
- Debug output showed identical numbers regardless of code changes
- Suggests the merge stage was sometimes reading stale alignment data

Adding `cuda.synchronize()` fixed this behavior immediately. It's definitely a "better safe than sorry" fix.

---

**Comment by Jamy-L** [comment]
I looked into the image size, and there is indeed a mismatch because the reference image is padded on the right and on the bottom before computing the pyramid. I did that to ensure that all the pixels in the reference image end up with an estimated optical flow (instead of giving a flow vector of 0,0 to the pixels on the edge). This may not be optimal but it attenuates the boundary effects slightly.

This wont have any effect on the coordinate system however, because the origin is on the top left. The moving images are not padded explicitly, because the padding is done on the fly (in the cuda kernels) whenever a search patch has to be read from them. I doubled check that the kernels systematically have `h, w = moving.shape` (and not ref.shape, which would be inaccurate due to padding). `h` and `w` are used in the kernels to detect out-of-bound pixels when reading the moving frame, so it is important to make sure we have the correct ones.

Regarding the burst, the issue I found is that it is extremely noisy and too dark, causing the alignment error to be indistinguishable from the noise, so the rejection mask doesn't work properly and most frames are accumulated together causing a blurry trail. I think this could be considered an extreme example, which would benefit from a different type of method

**Comment by sashko5l** [comment]
I also encountered this tiling issue (mild but noticeable, using both Nikon and Fuji X-trans Raws) in 1.3.2 that was corrected by changing from nearest to either bilinear or bicubic upscaling.
